File: Solutions_Medium/0102_BinaryTreeLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def _level_order_one_level(self, roots: List[TreeNode]) -> List[int]:
        return [root.val for root in roots]

    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        roots_on_level = [root]
        levels = []
        while len(roots_on_level) > 0:
            level = [root.val for root in roots_on_level]
            levels.append(level)
            new_roots = []
            for root in roots_on_level:
                if root.left is not None:
                    new_roots.append(root.left)
                if root.right is not None:
                    new_roots.append(root.right)
            roots_on_level = new_roots

        return levels

File: Solutions_Medium/0102_BinaryTreeLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode *root) {
        if (root == nullptr) {
            return {};
        }

        vector<TreeNode *> roots_on_level = {root};
        vector<vector<int>> levels;

        while (!roots_on_level.empty()) {
            vector<int> level;
            vector<TreeNode *> new_roots;

            for (TreeNode *root : roots_on_level) {
                level.push_back(root->val);

                if (root->left != nullptr) {
                    new_roots.push_back(root->left);
                }

                if (root->right != nullptr) {
                    new_roots.push_back(root->right);
                }
            }

            levels.push_back(level);
            roots_on_level = new_roots;
        }

        return levels;
    }
};